File: game-plugins/unreal/Source/SHAHCOIN/Private/SHAHCOINWallet.cpp
#include "SHAHCOINWallet.h"
#include "HttpModule.h"
#include "Interfaces/IHttpRequest.h"
#include "Interfaces/IHttpResponse.h"
#include "Dom/JsonObject.h"
#include "Serialization/JsonSerializer.h"
#include "Serialization/JsonWriter.h"
#include "Engine/Engine.h"

USHAHCOINWallet::USHAHCOINWallet()
{
    PrimaryComponentTick.bCanEverTick = true;
    
    // Initialize default values
    RPCUrl = TEXT("http://localhost:8368");
    Network = TEXT("mainnet");
    AutoConnect = true;
    BalanceUpdateInterval = 30.0f;
    
    // Initialize HTTP module
    HttpModule = &FHttpModule::Get();
}

void USHAHCOINWallet::BeginPlay()
{
    Super::BeginPlay();
    
    if (AutoConnect)
    {
        ConnectWallet();
    }
}

void USHAHCOINWallet::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
{
    Super::TickComponent(DeltaTime, TickType, ThisTickFunction);
    
    // Handle balance update timer
    if (IsConnected && BalanceUpdateInterval > 0.0f)
    {
        BalanceUpdateTimer += DeltaTime;
        if (BalanceUpdateTimer >= BalanceUpdateInterval)
        {
            BalanceUpdateTimer = 0.0f;
            GetBalance();
        }
    }
}

void USHAHCOINWallet::ConnectWallet()
{
    if (IsConnected)
    {
        UE_LOG(LogTemp, Warning, TEXT("SHAHCOIN Wallet already connected"));
        return;
    }
    
    UE_LOG(LogTemp, Log, TEXT("Connecting to SHAHCOIN wallet..."));
    
    // In a real implementation, this would connect to a wallet provider
    // For now, we'll simulate a connection
    FTimerHandle TimerHandle;
    GetWorld()->GetTimerManager().SetTimer(TimerHandle, [this]()
    {
        // Simulate getting wallet address
        CurrentAddress = FString::Printf(TEXT("SHAH%d"), FMath::RandRange(1000000000, 9999999999));
        IsConnected = true;
        
        // Broadcast connection event
        OnWalletConnected.Broadcast(CurrentAddress);
        
        UE_LOG(LogTemp, Log, TEXT("SHAHCOIN Wallet connected: %s"), *CurrentAddress);
        
        // Get initial balance
        GetBalance();
        
    }, 1.0f, false);
}

void USHAHCOINWallet::DisconnectWallet()
{
    if (!IsConnected)
    {
        return;
    }
    
    IsConnected = false;
    CurrentAddress = TEXT("");
    CurrentBalance = 0.0f;
    
    OnWalletDisconnected.Broadcast(TEXT(""));
    UE_LOG(LogTemp, Log, TEXT("SHAHCOIN Wallet disconnected"));
}

void USHAHCOINWallet::GetBalance()
{
    if (!IsConnected)
    {
        OnError.Broadcast(TEXT("Wallet not connected"));
        return;
    }
    
    // Create RPC request
    TSharedPtr<FJsonObject> RequestObject = MakeShareable(new FJsonObject);
    RequestObject->SetStringField(TEXT("jsonrpc"), TEXT("2.0"));
    RequestObject->SetNumberField(TEXT("id"), ++RequestId);
    RequestObject->SetStringField(TEXT("method"), TEXT("getbalance"));
    
    TArray<TSharedPtr<FJsonValue>> Params;
    Params.Add(MakeShareable(new FJsonValueString(CurrentAddress)));
    Params.Add(MakeShareable(new FJsonValueNumber(1)));
    RequestObject->SetArrayField(TEXT("params"), Params);
    
    // Convert to string
    FString RequestString;
    TSharedRef<TJsonWriter<>> Writer = TJsonWriterFactory<>::Create(&RequestString);
    FJsonSerializer::Serialize(RequestObject.ToSharedRef(), Writer);
    
    // Send HTTP request
    SendRPCRequest(RequestString, [this](const FString& Response)
    {
        ParseBalanceResponse(Response);
    });
}

void USHAHCOINWallet::SendSHAH(const FString& ToAddress, float Amount)
{
    if (!IsConnected)
    {
        OnError.Broadcast(TEXT("Wallet not connected"));
        return;
    }
    
    if (Amount <= 0.0f)
    {
        OnError.Broadcast(TEXT("Amount must be greater than 0"));
        return;
    }
    
    if (Amount > CurrentBalance)
    {
        OnError.Broadcast(TEXT("Insufficient balance"));
        return;
    }
    
    UE_LOG(LogTemp, Log, TEXT("Sending %.2f SHAH to %s"), Amount, *ToAddress);
    
    // In a real implementation, this would create and sign a transaction
    // For now, we'll simulate the transaction
    FTimerHandle TimerHandle;
    GetWorld()->GetTimerManager().SetTimer(TimerHandle, [this, ToAddress, Amount]()
    {
        // Simulate transaction
        FString TxId = FString::Printf(TEXT("tx_%d"), FMath::RandRange(1000000000, 9999999999));
        CurrentBalance -= Amount;
        
        OnTransactionSent.Broadcast(TxId);
        OnBalanceUpdated.Broadcast(CurrentBalance);
        
        UE_LOG(LogTemp, Log, TEXT("Transaction sent: %s, Amount: %.2f SHAH to %s"), 
               *TxId, Amount, *ToAddress);
        
    }, 2.0f, false);
}

void USHAHCOINWallet::GetNFTBalance()
{
    if (!IsConnected)
    {
        OnError.Broadcast(TEXT("Wallet not connected"));
        return;
    }
    
    UE_LOG(LogTemp, Log, TEXT("Getting NFT balance for wallet: %s"), *CurrentAddress);
    
    // In a real implementation, this would query the NFT registry
    // For now, we'll return mock data
    FTimerHandle TimerHandle;
    GetWorld()->GetTimerManager().SetTimer(TimerHandle, [this]()
    {
        // Simulate NFT data
        FNFTInfo NFTInfo;
        NFTInfo.TokenId = TEXT("1");
        NFTInfo.ContractAddress = TEXT("SHAH_NFT_001");
        NFTInfo.Name = TEXT("Game Achievement #1");
        NFTInfo.Description = TEXT("First victory in the arena");
        NFTInfo.Image = TEXT("https://example.com/nft1.png");
        
        // You can add an event to handle NFT data
        UE_LOG(LogTemp, Log, TEXT("Found NFT: %s"), *NFTInfo.Name);
        
    }, 0.5f, false);
}

void USHAHCOINWallet::SendRPCRequest(const FString& RequestData, TFunction<void(const FString&)> Callback)
{
    TSharedRef<IHttpRequest, ESPMode::ThreadSafe> Request = HttpModule->CreateRequest();
    Request->SetVerb(TEXT("POST"));
    Request->SetHeader(TEXT("Content-Type"), TEXT("application/json"));
    Request->SetURL(RPCUrl);
    Request->SetContentAsString(RequestData);
    
    Request->OnProcessRequestComplete().BindLambda([Callback](FHttpRequestPtr Request, FHttpResponsePtr Response, bool bSuccess)
    {
        if (bSuccess && Response.IsValid())
        {
            Callback(Response->GetContentAsString());
        }
        else
        {
            UE_LOG(LogTemp, Error, TEXT("RPC request failed"));
        }
    });
    
    Request->ProcessRequest();
}

void USHAHCOINWallet::ParseBalanceResponse(const FString& Response)
{
    TSharedPtr<FJsonObject> JsonObject;
    TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(Response);
    
    if (FJsonSerializer::Deserialize(Reader, JsonObject) && JsonObject.IsValid())
    {
        // Check for error
        if (JsonObject->HasField(TEXT("error")) && !JsonObject->GetObjectField(TEXT("error"))->IsEmpty())
        {
            FString ErrorMessage = JsonObject->GetObjectField(TEXT("error"))->GetStringField(TEXT("message"));
            OnError.Broadcast(FString::Printf(TEXT("RPC Error: %s"), *ErrorMessage));
            return;
        }
        
        // Parse result
        if (JsonObject->HasField(TEXT("result")))
        {
            float Balance = JsonObject->GetNumberField(TEXT("result"));
            CurrentBalance = Balance;
            OnBalanceUpdated.Broadcast(Balance);
        }
    }
    else
    {
        OnError.Broadcast(TEXT("Failed to parse RPC response"));
    }
}

// Blueprint exposed functions
void USHAHCOINWallet::ConnectWalletBP()
{
    ConnectWallet();
}

void USHAHCOINWallet::DisconnectWalletBP()
{
    DisconnectWallet();
}

void USHAHCOINWallet::GetBalanceBP()
{
    GetBalance();
}

void USHAHCOINWallet::SendSHAHBP(const FString& ToAddress, float Amount)
{
    SendSHAH(ToAddress, Amount);
}

void USHAHCOINWallet::GetNFTBalanceBP()
{
    GetNFTBalance();
}

// Getters for Blueprint
bool USHAHCOINWallet::GetIsConnected() const
{
    return IsConnected;
}

FString USHAHCOINWallet::GetCurrentAddress() const
{
    return CurrentAddress;
}

float USHAHCOINWallet::GetCurrentBalance() const
{
    return CurrentBalance;
}

File: game-plugins/unity/Runtime/Scripts/SHAHCOINWallet.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Networking;
using Newtonsoft.Json;

namespace SHAHCOIN
{
    /// <summary>
    /// SHAHCOIN Wallet Integration for Unity
    /// Handles wallet connection, balance queries, and basic transactions
    /// </summary>
    public class SHAHCOINWallet : MonoBehaviour
    {
        [Header("Network Configuration")]
        [SerializeField] private string rpcUrl = "http://localhost:8368";
        [SerializeField] private string network = "mainnet"; // mainnet, testnet, regtest
        
        [Header("Wallet Settings")]
        [SerializeField] private bool autoConnect = true;
        [SerializeField] private float balanceUpdateInterval = 30f;
        
        [Header("Events")]
        public UnityEvent<string> OnWalletConnected;
        public UnityEvent<string> OnWalletDisconnected;
        public UnityEvent<decimal> OnBalanceUpdated;
        public UnityEvent<string> OnTransactionSent;
        public UnityEvent<string> OnError;
        
        // Wallet state
        private bool isConnected = false;
        private string currentAddress = "";
        private decimal currentBalance = 0;
        private Coroutine balanceUpdateCoroutine;
        
        // RPC request ID counter
        private int requestId = 0;
        
        /// <summary>
        /// Wallet connection status
        /// </summary>
        public bool IsConnected => isConnected;
        
        /// <summary>
        /// Current wallet address
        /// </summary>
        public string CurrentAddress => currentAddress;
        
        /// <summary>
        /// Current wallet balance in SHAH
        /// </summary>
        public decimal CurrentBalance => currentBalance;
        
        private void Start()
        {
            if (autoConnect)
            {
                StartCoroutine(ConnectWallet());
            }
        }
        
        private void OnDestroy()
        {
            DisconnectWallet();
        }
        
        /// <summary>
        /// Connect to SHAHCOIN wallet
        /// </summary>
        public IEnumerator ConnectWallet()
        {
            try
            {
                // In a real implementation, this would connect to a wallet provider
                // For now, we'll simulate a connection
                yield return new WaitForSeconds(1f);
                
                // Simulate getting wallet address
                currentAddress = "SHAH" + UnityEngine.Random.Range(1000000000, 9999999999).ToString();
                isConnected = true;
                
                OnWalletConnected?.Invoke(currentAddress);
                
                // Start balance updates
                StartBalanceUpdates();
                
                Debug.Log($"SHAHCOIN Wallet connected: {currentAddress}");
            }
            catch (Exception e)
            {
                OnError?.Invoke($"Failed to connect wallet: {e.Message}");
                Debug.LogError($"SHAHCOIN Wallet connection failed: {e.Message}");
            }
        }
        
        /// <summary>
        /// Disconnect from wallet
        /// </summary>
        public void DisconnectWallet()
        {
            if (balanceUpdateCoroutine != null)
            {
                StopCoroutine(balanceUpdateCoroutine);
                balanceUpdateCoroutine = null;
            }
            
            isConnected = false;
            currentAddress = "";
            currentBalance = 0;
            
            OnWalletDisconnected?.Invoke("");
            Debug.Log("SHAHCOIN Wallet disconnected");
        }
        
        /// <summary>
        /// Get wallet balance
        /// </summary>
        public IEnumerator GetBalance()
        {
            if (!isConnected)
            {
                OnError?.Invoke("Wallet not connected");
                yield break;
            }
            
            var request = new RPCRequest
            {
                jsonrpc = "2.0",
                id = ++requestId,
                method = "getbalance",
                @params = new object[] { currentAddress, 1 }
            };
            
            yield return StartCoroutine(SendRPCRequest(request, (response) =>
            {
                if (response.error != null)
                {
                    OnError?.Invoke($"RPC Error: {response.error.message}");
                    return;
                }
                
                if (decimal.TryParse(response.result.ToString(), out decimal balance))
                {
                    currentBalance = balance;
                    OnBalanceUpdated?.Invoke(balance);
                }
            }));
        }
        
        /// <summary>
        /// Send SHAH to another address
        /// </summary>
        public IEnumerator SendSHAH(string toAddress, decimal amount)
        {
            if (!isConnected)
            {
                OnError?.Invoke("Wallet not connected");
                yield break;
            }
            
            if (amount <= 0)
            {
                OnError?.Invoke("Amount must be greater than 0");
                yield break;
            }
            
            if (amount > currentBalance)
            {
                OnError?.Invoke("Insufficient balance");
                yield break;
            }
            
            // In a real implementation, this would create and sign a transaction
            // For now, we'll simulate the transaction
            yield return new WaitForSeconds(2f);
            
            // Simulate transaction
            string txid = "tx_" + UnityEngine.Random.Range(1000000000, 9999999999).ToString();
            currentBalance -= amount;
            
            OnTransactionSent?.Invoke(txid);
            OnBalanceUpdated?.Invoke(currentBalance);
            
            Debug.Log($"Transaction sent: {txid}, Amount: {amount} SHAH to {toAddress}");
        }
        
        /// <summary>
        /// Get NFT balance for the connected wallet
        /// </summary>
        public IEnumerator GetNFTBalance()
        {
            if (!isConnected)
            {
                OnError?.Invoke("Wallet not connected");
                yield break;
            }
            
            // In a real implementation, this would query the NFT registry
            // For now, we'll return mock data
            yield return new WaitForSeconds(0.5f);
            
            var nfts = new List<NFTInfo>
            {
                new NFTInfo
                {
                    tokenId = "1",
                    contractAddress = "SHAH_NFT_001",
                    name = "Game Achievement #1",
                    description = "First victory in the arena",
                    image = "https://example.com/nft1.png"
                }
            };
            
            // You can add an event to handle NFT data
            Debug.Log($"Found {nfts.Count} NFTs in wallet");
        }
        
        /// <summary>
        /// Start automatic balance updates
        /// </summary>
        private void StartBalanceUpdates()
        {
            if (balanceUpdateCoroutine != null)
            {
                StopCoroutine(balanceUpdateCoroutine);
            }
            
            balanceUpdateCoroutine = StartCoroutine(BalanceUpdateLoop());
        }
        
        /// <summary>
        /// Balance update loop
        /// </summary>
        private IEnumerator BalanceUpdateLoop()
        {
            while (isConnected)
            {
                yield return StartCoroutine(GetBalance());
                yield return new WaitForSeconds(balanceUpdateInterval);
            }
        }
        
        /// <summary>
        /// Send RPC request to SHAHCOIN node
        /// </summary>
        private IEnumerator SendRPCRequest(RPCRequest request, System.Action<RPCResponse> callback)
        {
            string json = JsonConvert.SerializeObject(request);
            
            using (UnityWebRequest webRequest = new UnityWebRequest(rpcUrl, "POST"))
            {
                byte[] bodyRaw = System.Text.Encoding.UTF8.GetBytes(json);
                webRequest.uploadHandler = new UploadHandlerRaw(bodyRaw);
                webRequest.downloadHandler = new DownloadHandlerBuffer();
                webRequest.SetRequestHeader("Content-Type", "application/json");
                
                yield return webRequest.SendWebRequest();
                
                if (webRequest.result == UnityWebRequest.Result.Success)
                {
                    try
                    {
                        var response = JsonConvert.DeserializeObject<RPCResponse>(webRequest.downloadHandler.text);
                        callback?.Invoke(response);
                    }
                    catch (Exception e)
                    {
                        OnError?.Invoke($"Failed to parse RPC response: {e.Message}");
                    }
                }
                else
                {
                    OnError?.Invoke($"RPC request failed: {webRequest.error}");
                }
            }
        }
    }
    
    /// <summary>
    /// RPC request structure
    /// </summary>
    [Serializable]
    public class RPCRequest
    {
        public string jsonrpc;
        public int id;
        public string method;
        public object[] @params;
    }
    
    /// <summary>
    /// RPC response structure
    /// </summary>
    [Serializable]
    public class RPCResponse
    {
        public object result;
        public RPCError error;
        public int id;
    }
    
    /// <summary>
    /// RPC error structure
    /// </summary>
    [Serializable]
    public class RPCError
    {
        public int code;
        public string message;
    }
    
    /// <summary>
    /// NFT information structure
    /// </summary>
    [Serializable]
    public class NFTInfo
    {
        public string tokenId;
        public string contractAddress;
        public string name;
        public string description;
        public string image;
        public Dictionary<string, object> attributes;
    }
    
    /// <summary>
    /// Unity event for string parameters
    /// </summary>
    [Serializable]
    public class UnityEvent<T> : UnityEngine.Events.UnityEvent<T> { }
}
